File: src-tauri/src/architecture/domain/performance/process_memory_snapshot.rs
use serde::Serialize;

/// One line in the memory breakdown tooltip (status bar / diagnostics).
#[derive(Serialize, Clone, Debug)]
pub struct ProcessMemoryLineItem {
    /// OS process name, e.g. `vscode-rust-app`, `msedgewebview2`.
    pub name: String,
    pub pid: u32,
    /// Resident set / working set in megabytes.
    pub working_set_mb: f64,
    /// Windows private bytes when available; otherwise mirrors working set.
    pub private_mb: f64,
}

/// Accurate memory picture for the IDE process **tree**.
///
/// **Why not a single `memory_mb` field?**
/// Tauri hosts the UI in WebView2 child processes. Measuring only the Rust
/// host (~300 MB) under-reports what Task Manager shows when you expand the
/// app group (~800 MB–1 GB with chat, Monaco, emulators).
#[derive(Serialize, Clone, Debug)]
pub struct ProcessMemorySnapshot {
    /// Rust/Tauri host working set only.
    pub host_working_set_mb: f64,
    /// Sum of descendant processes (WebView2 renderer, GPU, utility).
    pub child_working_set_mb: f64,
    /// `host + children` — primary number shown in the status bar.
    pub total_working_set_mb: f64,
    /// Sum of private bytes (Windows); closer to Task Manager Details view.
    pub total_private_mb: f64,
    pub child_process_count: u32,
    pub cpu_usage_percent: f32,
    pub system_total_ram_gb: u64,
    pub system_available_ram_gb: u64,
    pub breakdown: Vec<ProcessMemoryLineItem>,
}

impl ProcessMemorySnapshot {
    /// Back-compat for older frontend code expecting `memory_mb`.
    pub fn legacy_memory_mb(&self) -> u64 {
        self.total_working_set_mb.round() as u64
    }
}

File: src/domain/performance/ProcessMemorySnapshot.ts
/**
 * Domain entity: live memory picture of the IDE process tree.
 *
 * WHY separate host vs children?
 * Tauri runs the UI in WebView2 child processes. Showing only the Rust host
 * (~300 MB) is misleading; Task Manager totals the whole tree (~800 MB–1 GB).
 */

export interface ProcessMemoryLineItem {
    name: string;
    pid: number;
    working_set_mb: number;
    private_mb: number;
}

export interface ProcessMemorySnapshot {
    host_working_set_mb: number;
    child_working_set_mb: number;
    total_working_set_mb: number;
    total_private_mb: number;
    child_process_count: number;
    cpu_usage_percent: number;
    system_total_ram_gb: number;
    system_available_ram_gb: number;
    breakdown: ProcessMemoryLineItem[];
}

/** Shape returned by `get_process_stats` Tauri command (includes legacy fields). */
export interface ProcessStatsDto {
    memory_mb: number;
    cpu_usage: number;
    total_ram_gb: number;
    available_ram_gb: number;
    snapshot: ProcessMemorySnapshot;
}

export function formatMemoryTooltip(s: ProcessMemorySnapshot): string {
    const lines = [
        `Total working set: ${s.total_working_set_mb.toFixed(0)} MB (host ${s.host_working_set_mb.toFixed(0)} + children ${s.child_working_set_mb.toFixed(0)})`,
        `Private bytes sum: ${s.total_private_mb.toFixed(0)} MB`,
        `Child processes: ${s.child_process_count}`,
        `CPU: ${s.cpu_usage_percent.toFixed(0)}%`,
        `System free: ${s.system_available_ram_gb} GB`,
    ];
    if (s.breakdown.length > 0) {
        lines.push('', 'Breakdown (top processes):');
        for (const row of s.breakdown.slice(0, 6)) {
            lines.push(`  ${row.name} (${row.pid}): ${row.working_set_mb.toFixed(0)} MB`);
        }
    }
    return lines.join('\n');
}
